gallery/tutorials/aspire_introduction.py

```diff
@@ -0,0 +1,662 @@
+#        CoordinateSource o-- BoxesCoordinateSource
+#        CoordinateSource o-- CentersCoordinateSource
+
+
+# %%
+# ``Simulation`` Class
+# ^^^^^^^^^^^^^^^^^^^^
+# Generating realistic synthetic data sources is a common task.  The
+# process of generating then projecting random rotations is integrated
+# into the `Simulation
+# <https://computationalcryoem.github.io/ASPIRE-Python/aspire.source.html#module-aspire.source.simulation>`_
+# class.  Using ``Simulation``, we can generate arbitrary numbers of
+# projections for use in experiments.  Then additional features are
+# introduced which allow us to create more realistic data sources.
+
+from aspire.source import Simulation
+
+# Total images in our source.
+num_imgs = 100
+
+# Generate a Simulation instance based on the original volume data.
+sim = Simulation(n=num_imgs, vols=v)
+# Display the first 10 images
+sim.images[:10].show()  # Hi Res
+
+# Repeat for the lower resolution (downsampled) volume v2.
+sim = Simulation(n=num_imgs, vols=v2)
+sim.images[:10].show()  # Lo Res
+
+# Note both of those simulations have the same rotations because they
+#   had the same seed by default, We recreate ``sim`` with a distinct
+#   seed to get different random samples (of rotations).
+sim = Simulation(n=num_imgs, vols=v2, seed=42)
+sim.images[:10].show()
+
+# We can also view the rotations used to create these projections.
+logging.info(sim.rotations)
+
+# %
+# Given any ``Source``, we can also take slices using typical slicing
+# syntax, or provide our own iterable of indices.
+
+sim_evens = sim[0::2]
+sim_odds = sim[1::2]
+
+# We can also generate random selections.
+# Shuffle indices then take the first 5.
+shuffled_inds = np.random.permutation(sim.n)[:5]
+sim_shuffled_subset = sim[shuffled_inds]
+
+# %
+# Underneath those slices, ASPIRE relies on ``IndexedSource``, which
+# we can also call direcly to remap indices.
+
+from aspire.source import IndexedSource
+
+sim_shuffled_subset = IndexedSource(sim, shuffled_inds)
+
+
+# %%
+# The ``noise`` Package
+# ---------------------
+# The `aspire.noise
+# <https://computationalcryoem.github.io/ASPIRE-Python/aspire.noise.html>`_
+# package contains several useful classes for generating and
+# estimating different types of noise.
+
+# %%
+# ``NoiseAdder``s
+# ^^^^^^^^^^^^^^^
+# ``NoiseAdder`` subclasses are used to add common or customized noise
+# to ``Simulation`` image generation pipelines.
+
+# %%
+# ``WhiteNoiseAdder``
+# """""""""""""""""""
+# ``WhiteNoiseAdder`` is the most common type of synthetic noise.
+
+from aspire.noise import WhiteNoiseAdder
+
+# Get the sample variance
+var = np.var(sim.images[:].asnumpy())
+logging.info(f"Sample Variance: {var}")
+target_noise_variance = 100.0 * var
+logging.info(f"Target Noise Variance: {target_noise_variance}")
+# Then create a NoiseAdder based on that variance.
+white_noise_adder = WhiteNoiseAdder(target_noise_variance)
+
+# %% We can customize Sources by adding stages to their generation
```

Comment:
`# %%` needs to be on its on line. The words on the same line are missing from the docs.

Reply:
nice